### DISCUSSION TITLE
Help importing a vsix

### DISCUSSION THREAD

**Discussion by ejfasting** [CodinGame/monaco-vscode-api#362] [Q&A] [answered]
Hello!

I had a discussion going on the monaco-languageclient-discussionboard about how you import a vsix (using vite): 
https://github.com/TypeFox/monaco-languageclient/discussions/597#discussioncomment-8697343 

I have set up a test-project, with vite, and removed everything besides my logic for importing the .vsix and showing a simple editor: 
https://github.com/ejfasting/MonacoTest/tree/main 

I get this when i run the import-statement at the top of main.ts.. :
Uncaught SyntaxError: malformed Unicode character escape sequence 

But from the other/old thread i see that @CGNonofr was able to import it without issues in the demo-project..
I'm not exactly sure what I'm doing wrong, but any input/feedback/help is much appreciated!

//eivind

**Comment by CGNonofr** [comment]
This library requires you to initialize the services using the `initialize` function from `vscode/services` with the service override you want

You should have a look at the demo

**Reply by ejfasting** [reply]
I see. My issue is primarily that its unable to even read the file:
![bilde](https://github.com/CodinGame/monaco-vscode-api/assets/60758190/f471b772-5399-4b32-a863-af120ea479ae)

I've looked at the demo, but unfortunately it didnt help much, and probably just got me more confused lol.. 
It also feels like kind of a maze to know whats what between monaco-vscode-api, monaco-languageclient and monaco-editor. 

Thank you for all your time! I think I have to give up on this and see if things have changed further down the line :)

**Reply by CGNonofr** [reply]
Are you using the rollup/vite plugin to load the file?

**Reply by ejfasting** [reply]
Yes. After a lot of testing i figured out the root cause of my issue had to do with the content of the vsix. Now it is imported as it should, and i can see the content of the file etc:
![image](https://github.com/CodinGame/monaco-vscode-api/assets/60758190/509f954d-142f-4bad-93d6-9bacad1d15ca)

Since server.js, which is the lsp-server, is bundled in the vsix it gets registered like this:
registerFileUrl('dist\server', 'data:application/javascript;base64,InVzZSBzdHJpY3QiO3ZhciBCUj1PY..... 

If i understand this correctly I then have to create the editor, register the servicesOverrides that i want, and somehow GET this registered file and launch it on the webworker (?).
If i had the .js-file directly in the monaco-project i would do this:
https://github.com/ejfasting/MonacoTest/blob/05ab15b3092f9b442e596f491f61dd394dfc5224/src/main.ts#L64 

Since the vsix actually contains the client and the server, is it possible to just import it and 'activate'?

**Reply by CGNonofr** [reply]
> Since server.js, which is the lsp-server, is bundled in the vsix it gets registered like this:
registerFileUrl('dist\server', 'data:application/javascript;base64,InVzZSBzdHJpY3QiO3ZhciBCUj1PY.....

Yes, in dev mode it's a base64 url, build in build mode, the asset is emitted

> If i understand this correctly I then have to create the editor, register the servicesOverrides that i want

Correct! (you just need to initialize the services BEFORE creating the editor)

> and somehow GET this registered file and launch it on the webworker (?).

It's already done in your extension so I'm not sure to understand your concerns

![image](https://github.com/CodinGame/monaco-vscode-api/assets/23257479/a9632d3b-69a7-4dd7-85f2-3bbcb62c223c)


> If i had the .js-file directly in the monaco-project i would do this:
https://github.com/ejfasting/MonacoTest/blob/05ab15b3092f9b442e596f491f61dd394dfc5224/src/main.ts#L64

You can still put that code in the extension code

> Since the vsix actually contains the client and the server, is it possible to just import it and 'activate'?

That's exactly what importing the vsix with the vsix-loader does


I think you're overthinking. You just need a VSCode web-compatible extension loaded via the vsix-loader and that's it
Just make sure:
- You've registered the required service overrides depending on the features you want
- You're provided a valid worker configuration to the extension service override (see https://github.com/CodinGame/monaco-vscode-api/blob/7864fdc200f6f8ee449056de624aeb7262e459cd/demo/src/setup.common.ts#L249 and https://github.com/CodinGame/monaco-vscode-api/blob/main/demo/src/tools/extHostWorker.ts) to enable to webworker extension host

**Reply by ejfasting** [reply]
Hm, ok. 
I know the extension is web-compatible, from testing it myself and by you in the demo-project. It loads up just fine, and gets imported. 

For starters I'm just adding these services, as I just want to load the vsix successfully:
`await initialize({
  ...getConfigurationServiceOverride(),
  ...getExtensionServiceOverride(workerConfig),
});
`
The workerConfig is basically a copy-paste of what you provided as an example. 

To set up the editor i do this:
`const editorOptions = {
    model: editor.createModel('const foo = () => 0;', languageId),
    automaticLayout: true
  };
  createConfiguredEditor(rootElement, editorOptions);`
  
  languageId here is jsfso, which is the same as the activationEvent in the .vsix. 
The editor gets added, with the text, but It does not seem to 'understand' its jsfso, which in turn should enable the extension (?). 
I added a short console.log to both the client and server, to validate that nothing happens. 

The console has other rows like this:
![image](https://github.com/CodinGame/monaco-vscode-api/assets/60758190/20977cd1-f873-414a-be83-fbb466be4762)
But the FAQ linked here says this could be because its served on file:// , but thats not the case when i use vite (?)

This probably goes outside of this discussion, and i appreciate your patience! :)

**Comment by CGNonofr** [comment]
> languageId here is jsfso, which is the same as the activationEvent in the .vsix. The editor gets added, with the text, but It does not seem to 'understand' its jsfso, which in turn should enable the extension (?). I added a short console.log to both the client and server, to validate that nothing happens.

Not sure what you are talking about: 
![image](https://github.com/CodinGame/monaco-vscode-api/assets/23257479/12063741-7832-48e3-acb8-138fd24309bd)


> The console has other rows like this: ![image](https://private-user-images.githubusercontent.com/60758190/311070833-20977cd1-f873-414a-be83-fbb466be4762.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk4ODkwMzMsIm5iZiI6MTcwOTg4ODczMywicGF0aCI6Ii82MDc1ODE5MC8zMTEwNzA4MzMtMjA5NzdjZDEtZjg3My00MTRhLWJlODMtZmJiNDY2YmU0NzYyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA4VDA5MDUzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFjYWUzZDNjM2M3YTQ4MjljNzYyZDY2MzQyN2MwNDY3Y2MwYmQxYWIzMTdmMTY3MTE4ODJiMGJjYTI4ZTMwOGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.s85Le_89E-5Pjh8_FmcKsbYaeREGi6t2ju6LITqFE_Q) But the FAQ linked here says this could be because its served on file:// , but thats not the case when i use vite (?)

Which FAQ? here it's about the editor worker that need to be configured (see https://github.com/CodinGame/monaco-vscode-api/blob/7864fdc200f6f8ee449056de624aeb7262e459cd/demo/src/setup.common.ts#L151-L167, you just need the first worker)

> This probably goes outside of this discussion, and i appreciate your patience! :)

No problem

**Reply by ejfasting** [reply]
Actually it turns out the webworker-stuff was not needed to make it work (when building for production). 
In dev-mode it fails to create the webworkers, but that part worked when you build for prod. 
That was probably what you were focusing on, and that is my bad for not being precise in the description of the problem. 

My missing piece turns out to be that I thought I didnt have to register a MonacoLanguageClient, because it was included in the vsix/vscode-extension. 
But after looking at the import of the vsix i see it only 'registerExtension', and the langium-example https://github.com/TypeFox/monaco-languageclient/blob/3979176106da5ff080381252b6a8873cd6e0c150/packages/examples/src/langium/statemachineClient.ts#L126 helped a lot. 

But, when you do the registerFileUrl it does not mean that Vite actually builds it on that path. I looked at the folder-structure of the dist and noticed that it puts the files from the vsix under: dist\assets\jsfso-client-0.0.1.vsix\ .. and creates a bogus name for the server.js:
![image](https://github.com/CodinGame/monaco-vscode-api/assets/60758190/44ac5e00-f1a9-42de-a90c-40edb0cc4056)

If i manually add that to the worker https://github.com/ejfasting/MonacoTest/blob/20b8001174843a1ebfa6ec5e3dcc5ec1f1342f8c/src/main.ts#L99 it shows the completionItems and outputs some logging-stuff that tells me the server launched. 

I would guess I should be able to fetchFileUrl or similar dynamically? Or somehow configure vite to output it in the correct place instead of extracting it into that sub-folder and append a random string to the name.

**Reply by CGNonofr** [reply]
Inside the extension, the url you provide to a Worker is transformed from the extension url to the "bogus" path (see https://github.com/microsoft/vscode/blob/27714e45c10df52ac2d9c6bf9b79e10fef00fb51/src/vs/workbench/api/worker/extensionHostWorker.ts#L96-L144)

The structure on the disk is not a concern at all

**Reply by ejfasting** [reply]
But, it does matter if i have to create the worker, in my monaco-extension, like this:
`const jsfsoWorkerUrl = new URL('assets/jsfso-client-0.0.1.vsix/**server-xyNSEqxM.js**', window.location.href).href;
const worker = new Worker(jsfsoWorkerUrl, {
  type: 'module',
  name: 'jsfso LS'
});`

This means that when server.js changes, because i add logic, I also need to edit the jsfsoWorkerUrl to have the new name of the file. 

What i was suggesting was that when this gets added through the vsix-import:
registerFileUrl('dist\server.js.map', 'data:application/json;base64,ewogICJ2ZXJzaW9u....

Then i should be able to do this, to not have to edit the filename every run:
const jsfsoWorkerUrl = new URL('dist/server.js', window.location.href).href;

I can probably use vite.config to do this, and tell it not to append the hash :)

**Reply by CGNonofr** [reply]
I'm really lost

What is a monaco-extension? is it different from the VSCode extension?

Are you trying to reference the worker from outside the extension? why?

**Reply by ejfasting** [reply]
Hehe, I understand, and my terminology is probably way off.. 
Take a look at the repo: https://github.com/ejfasting/MonacoTest/blob/main/src/main.ts 

I have still not made it work by simply importing the .vsix:
https://github.com/ejfasting/MonacoTest/blob/bbfb706811ccc0f1a0f6d6f0473ab9d0c369b311/src/main.ts#L3

adding ...getConfigurationServiceOverride:
https://github.com/ejfasting/MonacoTest/blob/bbfb706811ccc0f1a0f6d6f0473ab9d0c369b311/src/main.ts#L55

and uncomment these lines:
https://github.com/ejfasting/MonacoTest/blob/bbfb706811ccc0f1a0f6d6f0473ab9d0c369b311/src/main.ts#L34-L51

but it does work if i import the extension, then create a worker with the server.js and create a MonacoLanguageClient:
https://github.com/ejfasting/MonacoTest/blob/bbfb706811ccc0f1a0f6d6f0473ab9d0c369b311/src/main.ts#L96-L105

I'm all ears :)

**Reply by CGNonofr** [reply, marked as answer]
> Hehe, I understand, and my terminology is probably way off.. Take a look at the repo: https://github.com/ejfasting/MonacoTest/blob/main/src/main.ts
> 
> I have still not made it work by simply importing the .vsix: https://github.com/ejfasting/MonacoTest/blob/bbfb706811ccc0f1a0f6d6f0473ab9d0c369b311/src/main.ts#L3

> adding ...getConfigurationServiceOverride: https://github.com/ejfasting/MonacoTest/blob/bbfb706811ccc0f1a0f6d6f0473ab9d0c369b311/src/main.ts#L55

I don't think it will help you to make it work but it can't hurt

> and uncomment these lines: https://github.com/ejfasting/MonacoTest/blob/bbfb706811ccc0f1a0f6d6f0473ab9d0c369b311/src/main.ts#L34-L51

Same, you'll need it but the editor worker is able to run outside of the worker so not blocking

> but it does work if i import the extension, then create a worker with the server.js and create a MonacoLanguageClient: https://github.com/ejfasting/MonacoTest/blob/bbfb706811ccc0f1a0f6d6f0473ab9d0c369b311/src/main.ts#L96-L105

In that case, you're using the local extension host (allowed because you imported `vscode/localExtensionHost`)

> I'm all ears :)

Ok, here's the issues with your project:
- Vite have some issues with esbuild not supporting some feature in dev mode, to fix it, you should use the `@codingame/esbuild-import-meta-url-plugin` plugin (look at the demo here) and force it to optimize commonjs dependencies ('vscode-semver', 'vscode-marked')
- Your extension needs the worker extension host to run so you must provide the extension host worker configuration
- For your extension to trigger, the jsfso language should exist, your extension declare it but it requires the languages service override to be taken into account

note:
- monaco-languageclient is pulling old version of monaco-vscode-api and you don't seem to be needing it
- you don't need vscode/localExtensionHost

Here's the patch:

[fix.patch.zip](https://github.com/CodinGame/monaco-vscode-api/files/14539712/fix.patch.zip)

**Reply by ejfasting** [reply]
aha, that makes sense! Added the patch to the repo, but now says this:
![image](https://github.com/CodinGame/monaco-vscode-api/assets/60758190/f5e97f3f-52df-476b-a631-b9054bd5e5a5)

Fetch API cannot load extension-file://superoffice.jsfso-client/dist/client.js. URL scheme "extension-file" is not supported.

Thank you for all your help so far! This is getting really close :)

**Reply by CGNonofr** [reply]
I'm unable to reproduce your issue, what are you doing to produce that error?

![image](https://github.com/CodinGame/monaco-vscode-api/assets/23257479/973f0662-deef-4e28-a7a5-4e44dd4c5557)

**Reply by ejfasting** [reply]
Nothing special really:
1. clone the repo
2. npm install
3. npm run dev

Tried it on 2 different machines, just to make sure it wasnt just this environment. 
Looks to be related to that nativeFetch blocks a custom protocol.. -> https://github.com/electron/electron/issues/6749 

If you can do the 3 steps above without issues then this must be related to my environment and/or a lack of something outside of monaco.

**Reply by CGNonofr** [reply]
I've just checked another time and I confirm I don't have any issue with latest Chrome on Ubuntu

**Reply by ejfasting** [reply]
Ok, thank you :)

Guess its time to get a hold of a linux machine and see if it works there. It looks weird that it uses extension-file as the protocol, as far as I'm aware fetch requires http or https.. 

Thank you for amazing assistance on getting this up and running! There was a lot of moving parts i was unaware of, so I wouldnt have made it without your help :) 


//Eivind

**Reply by CGNonofr** [reply]
> Guess its time to get a hold of a linux machine and see if it works there. It looks weird that it uses extension-file as the protocol, as far as I'm aware fetch requires http or https..

fetch is not supposed to be called with the `extension-file` protocol as it's supposed to be replaced with the real url, but if that replacement failed for some reasons(escaping or formatting issue?), the extension-file url propagate to the fetch

What OS are you using? I suspect windows as there is some specific path replacement logic

> Thank you for amazing assistance on getting this up and running! There was a lot of moving parts i was unaware of, so I wouldnt have made it without your help :)

I see it as an opportunity to spot issues and better understand the VSCode codebase

**Reply by ejfasting** [reply]
Correct, I'm on windows 11 on both PC's, so i also believe there is something with the OS. 
If you are interested this is what is the errormsg i get in firefox:
message: "Activating extension 'SuperOffice.jsfso-client' failed: NetworkError when attempting to fetch resource.."
​I stepped into it and i see it tries to do the replace:
![image](https://github.com/CodinGame/monaco-vscode-api/assets/60758190/ed060fbe-50d6-4eb1-b021-86ef4eca4c18)
Havent been able to figure out exactly where it goes wrong though, will look into it tomorrow. 

This is btw on line 25 in extensionHost.Worker.js

**Reply by CGNonofr** [reply]
I've found the issue, it's related to the vsix loader

btw, do you know why your repo zip is detected as containing a virus by windows defender?

**Reply by CGNonofr** [reply]
https://github.com/CodinGame/monaco-vscode-api/pull/365

**Reply by ejfasting** [reply]
@CGNonofr oh, great, thanks!

No idea. Does it say which file it is? Tried scanning it here now (with windows defender) but nothing comes up.

**Reply by CGNonofr** [reply]
Just tried to download the zip from github and I'm not even able to open the file as the browser blocks it but there is no more information

BTW, it seems your vsix is really heavy because it contains both the web and node extension, so some BIG files as not used in the web version but still bundled. You should probably consider having a web dedicated extension with only required files

**Reply by ejfasting** [reply]
Weird.. I will do a better scan of the content. I'm able to download it through firefox an edge locally, and on 2 separate computers, and nothing shows up as suss.

I updated the .vsix in the project already. In the first version, which you probably are refering to, I totally forgot to add a .vscodeignore to the folder, so it included the whole node_modules folder. That was actually the issue with getting it to load initially, because something in the node_modules made the import crash. When i fixed that the .vsix shrunk a lot, and things make more sense. I was a little to focused on just getting the POC up :)
 
But thanks for the input! Excited to test out the bugfix!

**Reply by CGNonofr** [reply]
FYI

![image](https://github.com/CodinGame/monaco-vscode-api/assets/23257479/6219306a-6c0e-45cd-90e8-b21de7c18c0c)

![image](https://github.com/CodinGame/monaco-vscode-api/assets/23257479/1cb29d35-6fba-4b1b-8c23-164e041c2498)

**Reply by ejfasting** [reply]
Thank you for bringing it to my attention :) will have to figure out why its not detected locally, and not through windows defender OR malwarebytes (which at least should be able to detect it when scanning the zip).

**Reply by ejfasting** [reply]
Hi again @CGNonofr I've been digging into this. I was unable to get windows defender to give me the same error, so i formatted a PC and set up win11 from scratch there. I then ran windows defender on the zip, and it did indeed say it contained a virus. I then ran windows update, and updated windows defender, and then did a rescan.. and the error went away. Could this have been a false positive? What version of defender are you using, and if you have available updates could you install them and re-scan? Looks to me like a false positive..

**Reply by CGNonofr** [reply]
I was using a virtual machine probably out of date, so forget about it :+1:

**Reply by CGNonofr** [reply]
Can you give a try to the 3.1.0?

**Reply by ejfasting** [reply]
Looks to be working :) Thanks a lot!
